### PR TITLE
Android MapboxV10 ShapeSource clusterProperties and Kotlin port

### PR DESCRIPTION
**Pull request by VolkerLieber**
## Description

Implements #1745 for MapboxV10 (only for Android since @mfazekas already updated the ios side)

## Checklist

- [x] I have tested this on a device/simulator for each compatible OS
- [x] I updated the documentation with running `yarn generate` in the root folder
- [x] I mentioned this change in `CHANGELOG.md`
- [x] I updated the typings files (`index.d.ts`)
- [x] I added/ updated a sample (`/example`)

### CODE REVIEW

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.java [hunk 1]
```diff
@@ -109,6 +111,30 @@ public void setClusterMaxZoomLevel(RCTMGLShapeSource source, int clusterMaxZoom)
...
+        List<Map.Entry<String, ClusterPropertyEntry>> properties = new ArrayList<>();
+
+        ReadableMapKeySetIterator iterator = map.keySetIterator();
+        while (iterator.hasNextKey()) {
+            String name = iterator.nextKey();
+            ReadableArray expressions = map.getArray(name);
+
+            Expression operator;
+            if (expressions.getType(0) == ReadableType.Array) {
+                operator = ExpressionParser.from(expressions.getArray(0));
```
Comment (mfazekas) [inline android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.java:125]: Have you tested this for the case operator is a reduce expression?!
```js
{"sum": [["+", ["accumulated"], ["get", "sum"]], ["get", "scalerank"]]}
```
In this case would mapbox accept `["+", ["accumulated"], ["get", "sum"]]` as an operator?!

It should accept the simple case:
```js
{"sum": ["+", ["get", "scalerank"]]}
```



https://github.com/mapbox/mapbox-maps-ios/blob/c61ccd41febfb2df2ed35e8155a566f37b8163b6/Sources/MapboxMaps/Style/Generated/Sources/GeoJSONSource.swift#L39-L45

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/utils/ClusterPropertyEntry.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.mapbox.rctmgl.utils;
+
+
+import com.mapbox.maps.extension.style.expressions.generated.Expression;
+
+public class ClusterPropertyEntry {
+    public Expression operator;
+    public Expression mapping;
+
+    public ClusterPropertyEntry(Expression _operator, Expression _mapping) {
```
Comment (mfazekas) [inline android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/utils/ClusterPropertyEntry.java:10]: Wouldn't make sense to add the `ReadableArray` conversion here?! What do you think?!

**Review by mfazekas — COMMENTED** [review]
👍 awesome, thanks much for the contribution

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/utils/ClusterPropertyEntry.java (hunk 2 above)
Comment (VolkerLieber) [reply android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/utils/ClusterPropertyEntry.java]: Yes, you are right. To do it correctly, I also implemented a correct type switch

**Comment by VolkerLieber** [comment]
Please give me a few minutes until I can update all of the changes

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.java (hunk 1 above)
Comment (VolkerLieber) [reply android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.java:125]: `{"sum": [["+", ["accumulated"], ["get", "sum"]], ["get", "scalerank"]]}`
`{"sum": ["+", ["get", "scalerank"]]}`
works, but although 
`["+", ["accumulated"], ["get", "sum"]]`
could be implemented like this
```
package com.mapbox.rctmgl.utils

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableType
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.mapbox.maps.extension.style.expressions.generated.Expression
import java.util.*

class ClusterPropertyEntry(expressions: ReadableArray) {
    @JvmField
    var expression: Expression

    init {
        val operator: String = when (expressions.getType(0)) {
            ReadableType.Array -> Gson().toJson(ConvertUtils.toJsonArray(expressions.getArray(0)))
            ReadableType.Map -> Gson().toJson(ConvertUtils.typedToJsonElement(expressions.getMap(0)) as JsonArray)
            ReadableType.Boolean -> String.format(Locale.ENGLISH, "%b", expressions.getBoolean(0));
            ReadableType.Number -> String.format(
                Locale.ENGLISH,
                "%d",
                expressions.getDouble(0)
            ); // All JS numbers are doubles
            else -> expressions.getString(0) // String, Null
        }

        val builder: Expression.ExpressionBuilder = Expression.ExpressionBuilder(operator)

        for (iExp in 1 until expressions.size()) {
            builder.addArgument(
                when (expressions.getType(iExp)) {
                    ReadableType.Array -> ExpressionParser.from(expressions.getArray(iExp))
                    ReadableType.Map -> ExpressionParser.from(expressions.getMap(iExp))
                    ReadableType.Boolean -> Expression.literal(expressions.getBoolean(iExp))
                    ReadableType.Number ->                 // All JS numbers are doubles
                        Expression.literal(expressions.getDouble(iExp))
                    else ->                 // String, Null
                        Expression.literal(expressions.getString(iExp))
                }!!
            )
        }

        expression = builder.build()
    }
}
```

it's currently not supported by Mapbox
`W/System.err: com.mapbox.maps.MapboxStyleException: Add source failed: GeoJSON source clusterProperties member must be an array with length of 2`

**Comment by VolkerLieber** [comment]
Would you prefer the Kotlin ClusterPropertyEntry implementation supporting multiple expression arguments, even though they are not supported by Mapbox at the moment, to be safe for future updates or do you prefer the Java implementation in my pull request?

**Comment by mfazekas** [comment]
> Would you prefer the Kotlin ClusterPropertyEntry implementation supporting multiple expression arguments (even though they are not supported by Mapbox at the moment) to be safe for future updates or do you prefer the Java implementation in my pull request?

We're migrating code to koltin, so it would make sense to implement  ClusterPropertyEntry in kotlin.

**Comment by VolkerLieber** [comment]
Ok, something is not working correctly in the Kotlin version. I have some other things to finish today as well, but I'm going to fix it first thing tomorrow

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.kt [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        source.setClusterRadius(radius.toLong())
+    }
+
+    @ReactProp(name = "clusterMaxZoomLevel")
+    fun setClusterMaxZoomLevel(source: RCTMGLShapeSource, clusterMaxZoom: Int) {
+        source.setClusterMaxZoom(clusterMaxZoom.toLong())
+    }
+
+    @ReactProp(name = "clusterProperties")
+    fun setClusterProperties(source: RCTMGLShapeSource, map: ReadableMap) {
```
Comment (VolkerLieber) [inline android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.kt:86]: Parse clusterProperties to HashMap (which can be directly used for setting the clusterProperties in Mapbox)

Comment (VolkerLieber) [reply android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSourceManager.kt:86]: And I ported ShapeSource, ShapeSourceManager to Kotlin

**Comment by VolkerLieber** [comment]
@mfazekas Please review again if you have the time

**Review by mfazekas — APPROVED** [review]

**Pull request merged**